:seedling: Add a design document to filter cache by selectors (#1419)
kubernetes-sigs/controller-runtime: designs/use-selectors-at-cache.md

...
}

```

Incrase `cache.Options` as follow:

```golang
type Options struct {
	Scheme                  *runtime.Scheme
	Mapper                  meta.RESTMapper
	Resync                  *time.Duration
	Namespace               string
	SelectorByResource selector.SelectorsByGroupResource
}
```

Add new builder function that will return a cache constructor that will return
a cache filtered by selectors

```golang
// BuilderWithSelectors returns a Cache constructor that will filter
// contents using selectors
// WARNING: filtered out resources are not returned.
func BuilderWithSelectors(selectors selector.SelectorsByGroupResource) NewCacheFunc {
...
}
```

is passed to informer's ListWatch and add the filtering option:

```golang

# At pkg/cache/internal/informers_map.go

ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
            ip.selectors.FindByGR(mapping.Resource).FillInListOpts(&opts)
...
```

Here is a PR with the implementatin at the `pkg/cache` part [3]

## Example

User will override `NewCache` function to make clear that they know exactly the
implications of using a different cache than the default one

```golang

>  ctrl.Options.NewCache = cache.BuilderWithSelectors(

Reviewer: General remark/ask: Would this proposal support setting up different caches for the same resource but for different controllers?
We have use cases for leveraging different optimized watches (e.g. with mutually exclusive label selectors / different namespaces) in different controllers. Would be nice, if we could build this right into this optimized cache.

Author: We have one cache per Manager, to do so you would need different manager (maybe running at different containers) or running multiple managers in the same container (I suppose this can be problematic).

Making a cache per controller is out of the scope of this design document.

Reviewer 2: @timebertt you can already do "one cache per controller" today by using `source.KindWithCache`. It is not particularly ergonomic, but it works fine

Reviewer: Ah, right. Missed that option. 
Thanks for pointing it out 👍

                            selector.SelectorsByGroupResource{

> 							    {Group: "", Resource: "nodes"}: {

Reviewer: I would probably go for `client.Object` instead of `schema.GroupResource` for configuring different caches (similar to how it is done in [`NewDelegatingClient`](https://github.com/kubernetes-sigs/controller-runtime/blob/eb0f1d4aa81794a38206965be25d118af5e18afc/pkg/client/split.go#L44).
It's more safe and consistent.
```suggestion
							    &corev1.Node: {
```

Author: Make sense, I there is a `apiutil.GVKForObject(obj, in.Client.Scheme())` to compose the GVK from a client.Object.

Reviewer: The map for lookup can be built up once before hand, just like in the `delegatingReader`:
https://github.com/kubernetes-sigs/controller-runtime/blob/eb0f1d4aa81794a38206965be25d118af5e18afc/pkg/client/split.go#L46-L52

Anyways, this can be discussed in detail during the implementation review.

Reviewer 2: I agree on the `client.Object` remark

Author: Done

Filter cache ListWatch using selectors
===================
## Motivation

> The controller-runtime subscribe to changes using a cache mechanism that get
> updated using kubernetes informers, those informers implement a `ListWatch`
> that specify the namespace and type specified at controller builder time.

Reviewer 2: ```suggestion
Controller-Runtime controllers use a cache to subscribe to events from Kubernetes objects and to read those objects more efficiently by avoiding to call out to the API. This cache is backed by Kubernetes informers.
```

Author: Done

> So the only possible filtering of that cache is namespace and resource type,
> this means that if some user is interested only on Reconcile some resource with
> specific fields it will still cache all the instances on that type increasing

Reviewer 2: ```suggestion
The only way to filter this cache is by namespace and resource type. In cases where a controller is only interested in a small subset of objects (for example all pods on a node), this might end up not being efficient enough.
```

Author: Done

the memory and CPU load. This also get worse in case controller-runtime is

> deployed as a daemonset instead of a deployment since all the nodes with

Reviewer 2: I don't think how you deploy your controller is particularly relevant, the relevant aspect from your example is that your controller instance is only interested in a small subset of all objects that can not be filtered via namespace or type, right?

Author: Sure, removing it, your previous correction is good enough.

daemonset's pods will expect to receive notification of all the instances and
that can end with scalation problems depending on the type or resource it's
watching.

> As side effect of this proposal the request that does not match the filtered
> will not be returned, so users has to be warned that if this filtering
> is used they know it and they are careful with the request they are doing.

Reviewer 2: ```suggestion
Requests to a client backed by a filtered cache for objects that do not match the filter will never return anything, so
we need to make sure that we properly warn users to only use this when they are sure they know what they are doing.
```

Author: Done

> As a side note this proposal sidesteps some of the long standing issues with
> other [solutions](https://github.com/kubernetes-sigs/controller-runtime/issues/244#issuecomment-451336202)
> like problematic interactions between controllers or the need for a complex
> refactorings by adding the possibility for users to define the ListWatch
> selectors but with an API that will be used by users that know what are the
> implications.

Reviewer 2: ```suggestion
This proposal sidesteps the issue of "How to we plug this into the cache-backed client so that users get feedback when they request something that is not matching the caches filter" by only implementing the filter logic in the cache
package. This allows advanced users to combine a filtered cache with the already existing `NewCacheFunc` option in the manager and cluster package, while simultaneously hiding it from newer users that might not be aware of the implications and the associated foot-shoot potential.
```

Author: Done

> The alternative that can be done with the current controller-runtme is
> to implement a custom cache that do the filtering since cache is pluggable
> into the controller-runtime [1]

Reviewer 2: ```suggestion
The only alternative today to get a filtered cache with controller-runtime is to build it out-of tree. Because such a cache would mostly copy the existing cache and add just some options, this is not great for consumers.
```

Author: Done

deployed as a daemonset instead of a deployment since all the nodes with
daemonset's pods will expect to receive notification of all the instances and
that can end with scalation problems depending on the type or resource it's
watching.

As side effect of this proposal the request that does not match the filtered
will not be returned, so users has to be warned that if this filtering
is used they know it and they are careful with the request they are doing.

As a side note this proposal sidesteps some of the long standing issues with
other [solutions](https://github.com/kubernetes-sigs/controller-runtime/issues/244#issuecomment-451336202)
like problematic interactions between controllers or the need for a complex
refactorings by adding the possibility for users to define the ListWatch
selectors but with an API that will be used by users that know what are the
implications.

The alternative that can be done with the current controller-runtme is
to implement a custom cache that do the filtering since cache is pluggable
into the controller-runtime [1]

This proposal is related to the following issue [2]

## Proposal

> Add a new selector code at `pkg/cache/selector/selector.go` with common structs

Reviewer 2: Why a new `selector` package rather than putting this into the existing `cache` package?

Author: `pkg/cache/internal` is importing it, at golang you cannot import "parent" packages.

Reviewer 2: So define it in pkg/cache/internal and alias it from pkg/cache?

Author: Make sense, done.

and helpers
```golang
package selector

...

// SelectorsByGroupResource associate a GroupResource to a field/label selector
type SelectorsByGroupResource map[schema.GroupResource]Selector

// Selector specify the label/field selector to fill in ListOptions
type Selector struct {
	Label labels.Selector
	Field fields.Selector
}

// FillInListOpts fill in ListOptions LabelSelector and FieldSelector if needed
func (s Selector) FillInListOpts(listOpts *metav1.ListOptions) {
...
}

// FindByGR use the GVR group and resource to find the field/label selector
func (s SelectorsByGroupResource) FindByGR(gvr schema.GroupVersionResource) Selector {
...
}

```

> Incrase `cache.Options` as follow:

Reviewer 2: Nit:

```suggestion
Extend `cache.Options` as follows:
```

Author: Done

```golang
type Options struct {
	Scheme                  *runtime.Scheme
	Mapper                  meta.RESTMapper
	Resync                  *time.Duration
	Namespace               string
	SelectorByResource selector.SelectorsByGroupResource
}
```

Add new builder function that will return a cache constructor that will return
a cache filtered by selectors

```golang
// BuilderWithSelectors returns a Cache constructor that will filter
// contents using selectors
// WARNING: filtered out resources are not returned.

> func BuilderWithSelectors(selectors selector.SelectorsByGroupResource) NewCacheFunc {

Reviewer 2: Why just selectors and not `Options`?

the filter will never return anything, so we need to make sure that we properly
warn users to only use this when they are sure they know what they are doing.

This proposal sidesteps the issue of "How to we plug this into the cache-backed
client so that users get feedback when they request something that is
not matching the caches filter" by only implementing the filter logic in the
cache package. This allows advanced users to combine a filtered cache with the
already existing `NewCacheFunc` option in the manager and cluster package,
while simultaneously hiding it from newer users that might not be aware of the
implications and the associated foot-shoot potential.

The only alternative today to get a filtered cache with controller-runtime is
to build it out-of tree. Because such a cache would mostly copy the existing
cache and add just some options, this is not great for consumers.

This proposal is related to the following issue [2]

## Proposal

Add a new selector code at `pkg/cache/internal/selector.go` with common structs
and helpers
```golang
package internal

...

// SelectorsByObject associate a client.Object to a field/label selector

> type SelectorsByObject map[client.Object]Selector

Reviewer 3: This should be restricted to `runtime.Object` so users don't get the idea that you can filter by object name/namespace
```suggestion
type SelectorsByObject map[runtime.Object]Selector
```

Author: Done

Reviewer 2: @estroz why? It is a map, the values can not access the keys. Making this `runtime.Object` means we have to deal with stuff like ppl putting List types in there

Author: @alvaroaleman don't we have the same problem with client.Object since it's embedding runtime.Object ?

Reviewer 3: @alvaroaleman it may be confusing to users to have all these object fields exposed by the `metav1.Object`, since they may assume the key can actually apply its own selector values, ex. by `SetName()`. Instead it would be better to constrain the contract to just `GroupVersionKind()`.

There can be some internal validation step to make sure the object is not a list type.

Reviewer 2: Using GVK is inconvenient. Using the smaller interface (runtime.Object) doesn't prevent ppl from putting in objects that fulfill a lager interface (client.Object). It does however keep them from putting something like List types in there and makes the compiler verify that rather than getting a runtime error. Also, I think the idea that ppl might think the map key has  any further relevance for filtering when the value is a `Selector` struct seems a bit far fetched to me.

We also have plenty of prior art where we use client.Object to only derive the type, e.G. in source.Kind or the clients delete method.

Reviewer 3: You could similarly argue that someone would not try to map a selector to a list type because that's not the type they add to `For()`, `Owns()`, or `Watches()`.

However since there is prior art I am fine with `client.Object`. There can always be another selector map type added later for `runtime.Object` without breakage if we see bugs filed because of confusion. It's not worth worrying about too much now.

Author: Going back to `client.Object` as proposed originally.

> // SelectorsByObject associate a GroupResource to a field/label selector
> type SelectorsByGVK map[schema.GVK]Selector

Reviewer 3: Nit
```suggestion
// SelectorsByObject associate a GroupVersionKind to a field/label selector
type SelectorsByGVK map[schema.GroupVersionKind]Selector
```

Author: Done

Reviewer 2: @estroz why? It is a map, the values can not access the keys. Making this `runtime.Object` means we have to deal with stuff like ppl putting List types in there

// Selector specify the label/field selector to fill in ListOptions
type Selector struct {
	Label labels.Selector
	Field fields.Selector
}

// FillInListOpts fill in ListOptions LabelSelector and FieldSelector if needed

> func (s Selector) FillInListOpts(listOpts *metav1.ListOptions) {

Reviewer 3: How about
```suggestion
func (s Selector) ApplyToList(listOpts *client.ListOptions) {
```
so it implements `client.ListOption`

Author: Done

...
}

Add a type alias to `pkg/cache/cache.go` to internal

```golang

type SelectorsByObject internal.SelectorsByObject

```

Extend `cache.Options` as follows:

```golang
type Options struct {
	Scheme            *runtime.Scheme
	Mapper            meta.RESTMapper
	Resync            *time.Duration
	Namespace         string

> 	SelectorsByObject SelectorsByObject

Reviewer 3: It would be nice to generalize this to something like 
```suggestion
	Selectors SelectorSet
```
Where `SelectorSet` is any set of selectors that implements
```go
type SelectorSet interface {
	client.ListOption
	Select()
}
```
where `Select()` is just an identifier method so users don't pass in other `client.ListOption`s.

Author: Done

Author: Do you mean ?

```golang
type SelectorSet interface {
	Select(runtime.Object) client.ListOption
}
```

Author: We have apply options per GVK so it doe not make sense to implement ListOption at the map level.

Reviewer 3: Sure, get rid of `client.ListOption` from that interface. Regardless, I may want to construct a GVK myself instead of using a typed object or `Unstructured`, so why not also expose `SelectorsByGVK`?

Author: As we have previously stated we are going to expose the map key as `client.Object`.
